### ISSUE TITLE
Video use in Jamtaba discussion

### ISSUE THREAD

**Issue by jonjamcam** [elieserdejesus/JamTaba#1073]
Video possibly out of sync (Tested on Jamtaba 2.1.4 -not h264 codec-)

I've been paying a lot of attention to the video and how is it in sync with channel 1 (which is in most cases the one used). What I've noticed is that video is **half a bpi interval early**.

I was expecting a whole bpi interval, but it looks to me that the difference is half a whole interval. I saw this behavior with trashcan's cam, so I did and experiment with my cam asking torben and in this case the cam was in perfect sync.

So atm I'm not conclusive about it. could be many things

**Comment by elieserdejesus** [comment]
> I was expecting a whole bpi interval

This is the expected result.

Maybe trashcan's cam is xmiting high resolution video (even in the cam lower resolution) and dropping some video frames. 

I don't have a good cam to test this scenario, maybe this can be testable using some cam software.

**Comment by jonjamcam** [comment]
> Maybe trashcan's cam is xmiting high resolution video (even using the cam lower resolution) and dropping some video frames.

I believe frame dropping may be the cause too. We'll see

**Comment by jonjamcam** [comment]
Some remarks make by andy about video use in Jamtaba:

![image](https://user-images.githubusercontent.com/15310433/49323303-72d27380-f500-11e8-93e4-7b5b5eff4715.png)

**Comment by elieserdejesus** [comment]
Tested today using 2 scenarios:
1 - **using my laptop webcam**: Jamtaba is choosing a low resolution (320 x 240) and works as expected when connected in ninbot. No disconnections!

2 - **Using magic cam sofware**: I forced magic cam to highest resolution (1680 x 1050) and I'm disconnected in the first interval, or the second (using 16 BPI). JamTaba was slow, and I'm using a good processor and the CPU usage is 34%. 

**My conclusion at moment**: Some heavy image processing is doing in Qt thread, this is slowing the UI and possibility the network code (managed by Qt too). My hypothesys is network is unresponsive because Qt is processing (I think is resizing) the webcam frames. I will run other tests to check this issue.

**Comment by elieserdejesus** [comment]
@jonjamcam , can you test the PR https://github.com/elieserdejesus/JamTaba/pull/1163 ??

I tested here simulating a big webcam and is working fine. Locally the webcam image is big, but a 320z240 version is xmited to server.

Solved the disconnection problems in my tests.

**Comment by jonjamcam** [comment]
>  can you test the PR #1163 ?

Working here @elieserdejesus . I see size is fixed to 320x240, but I see little difference. No disconnections though.

**Comment by elieserdejesus** [comment]
> Was this the original problem -audio disconnection-?

No @jonjamcam , the original problem is a server disconnection.

**Comment by jonjamcam** [comment]
>the original problem is a server disconnection.

I remember server disconnecting some time ago, but I attributed it to slow connection  https://github.com/elieserdejesus/JamTaba/issues/1021

 But after upgrading to 4g no more problems. This was using previous mp4 codec. 

I must say I haven't experienced server disconnections using the new codec. Only some occasional muting in the audio and sometimes frame drop.

I'll do some more testing @elieserdejesus

**Comment by elieserdejesus** [comment]
> I must say I haven't experienced server disconnections using the new codec

@jonjamcam , the problem it's related with the camera resolution. I think the PR https://github.com/elieserdejesus/JamTaba/pull/1163 is solving this issue. Can you test the PR?

**Comment by jonjamcam** [comment]
I did some tests with both internal camera at 320x240  and droidcam at 640x480. Working fine @elieserdejesus

**Comment by elieserdejesus** [comment]
thanks @jonjamcam , I will merge the PR!  Can we close this issue? What you think?

**Comment by jonjamcam** [comment]
>Can we close this issue? What you think?

Yes. I agree. It's solved.

**Comment by jonjamcam** [comment]
testing with a typical 6 user ninbot server with @andy and I using video and both got disconnected from the server some minutes after turning video on

**Comment by elieserdejesus** [comment]
I think the CBR (Constant Bit Rate) mentioned by Andy make sense. I will reopen the issue.

**Comment by elieserdejesus** [comment]
Maybe this will be usefull: https://ffmpeg.zeranoe.com/forum/viewtopic.php?t=2417

**Comment by jonjamcam** [comment]
Did some real life testing on ninbot and ninjamer with trashcanhead. Some conclusions atm:

1.- h264 encoding (using 2.1.10) makes the client crash in both ninbot and ninjamer servers.

2.- previous encoder (mp4?) Works fine in both ninjamer and ninbot. We both used jamtaba 2.1.4 for this test.

BUT.... During this last test with 2.1.4 I saw some reaninjam clients crash several times. Not sure if it was because of the video usage.

**Comment by jonjamcam** [comment]
Second test. no problems with video. we had 3 users to jam with 2.1.4 on ninbot server (trashcanhead, Golstrider and me) along with 4 other jamtaba and reaninjam for some hours.

**Comment by Trashcanhead1111** [comment]
I just tried using Many cam and setting camera to 320x240 with version 2.1.10 still had Crash after one interval. Also tried different cbr and vbr bitrate. unsure what to try. it kicks after one interval it seems. Could it be toms server? Running windows 10 also here.

**Comment by elieserdejesus** [comment]
I will check the video feature in the next "round".

**Comment by Trashcanhead1111** [comment]
Taiwan user is using mac and video working fine with 2.1.10 so its only windows version .

**Comment by jonjamcam** [comment]
Testing 2.0.11 video feature. What I find is Jamtaba is closing the connection if I turn the camera on sfter I join the server.

@elieserdejesus is it just me? can you confirm?

If I turn the camera on BEFORE I join a server, seems to work fine at least in 1 to 1 tests in my own computer

**Comment by elieserdejesus** [comment]
@jonjamcam , teste here using 2 jamtabas connected in same server, private server and in ninbot, I have no problems. I tested turn camera on just after connect (still preparing) and after preparing (transmiting). No errors.

Can you share more details?:
1 - bpi and bpm of server
2 - When you turn on the camera? In preparing state or in transmiting state?
3 - In the log file what jamtaba is reporting as the camera resolution? 
4 - What is the video preview size? (the value is closest to the log line showing the camera resolution) 
5 - Webcam? Droidcam?

**Comment by arnaudlecam** [comment]
Hi,
Video is causing an unexpected closure (with segfault) for me too.
Here are my final debug messages:
```
default.CRITICAL  [01:14:37:596] Could not find encoder for  indeo3 [, line 0]
default.CRITICAL  [01:14:37:701] Configurator::signalHandler signal: 11 [, line 0]
default.CRITICAL  [01:14:37:703] Configurator::signalHandler signal: 11 [, line 0]
malloc_consolidate(): invalid chunk size
free(): corrupted unsorted chunks
```
Somtimes followed by:
```
default.WARNING   [01:38:21:612] QObject::~QObject: Timers cannot be stopped from another thread [, line 0]
pure virtual method called
default.CRITICAL  [] Configurator::signalHandler signal: 6 [, line 0
```
1. bpi and bpm of server
it looks totaly intependant: crashing systematically.
2. When you turn on the camera? In preparing state or in transmiting state?
each time I try to use it in a ninjam room and only when my client tries to transmit.  Out of the ninjam room, or while no preparing/transmitting: there is no problem.
3. In the log file what jamtaba is reporting as the camera resolution?
`default.CRITICAL  [01:38:37:281] Camera resolutions list is empty! [, line 0]`
(it appears each time I clic on the webam's icon to activate)
4. What is the video preview size? (the value is closest to the log line showing the camera resolution)
I don't know.
5. Webcam? Droidcam?
I don't know the difference.

**Comment by jonjamcam** [comment]
my recent test was:

1.- open Jamtaba standalone
2.- connect to public server (4 clients online), some reaper users, some jamtaba. No video used.
3.- Turn on camera.
4.- after 2 bpi intervals, my client is disconnected. Jamtaba is NOT crashing, just disconnecting.
5.- closing Jamtba

clean log file:

[log.txt](https://github.com/elieserdejesus/JamTaba/files/4453352/log.txt)

**Comment by jonjamcam** [comment]
test 1:

testing using 2.0.11 compiled with qt 5.6.3 (my side):

![image](https://user-images.githubusercontent.com/15310433/78849497-a53c8c80-79e2-11ea-9277-4fb70c7abed6.png)

jeff and trashcanhead are using 2.1.4 for video stability.

No problems atm. I turned on my camera before entering the room.

Test2:

**Comment by jonjamcam** [comment]
bad news. Trashcanhead and jeff tested the new 2.0.11 beta and the problem still exist for them. In my system as far as I turn the camera on before entering the room it works, but not for them.

back to drawing board

**Comment by elieserdejesus** [comment]
@arnaudlecam , your situation is totally unexpected. 

1 - From JamTaba is not finding the encoder, but the encoder is inside a static library (libavcodec) compiled in the executable. Are you compiling JamTaba or using a released version?

2 - Your camera is not returning a resolutions list, very weird too. JamTaba is showing your camera preview in the bottom left?

3 - Are you in Linux, Mac or Windows?

**Comment by elieserdejesus** [comment]
@jonjamcam , thanks for the hard work testing and collecting the data.

**Comment by arnaudlecam** [comment]
Hi @elieserdejesus ,
1. your question reminds me I had errors when I compiled for the first time, so I linked a few static libs with mine and it worked. Now I understand my mystake, thank you ! My first step will be to copy my local libs to statics, before coming back ;
2. preview is ok ;
3. here is my 'uname -a' under archlinux : `Linux 4.19.106-rt45-2-rt-lts #1 SMP PREEMPT RT Mon, 30 Mar 2020 17:02:56 +0000 x86_64 GNU/Linux` ;

**Comment by arnaudlecam** [comment]
So, I re-arranged my static libs (30 Mar 2020 upgraded), linking to all present /usr/lib and /usr/include, and recompiled.
Let's see a test, when I connect 2 users on a private distant server:
![IMG_20200410_005233954](https://user-images.githubusercontent.com/12881960/78948411-81537680-7ac8-11ea-8fe4-0befd02bd324.jpg)
It's OK now !
but I still have the `Camera resolutions list is empty! [, line 0]`
I'll try to recompile with [pre compiled libs](https://www.dropbox.com/s/qckwsmaqlditwpb/JamTaba-static-libs.zip?dl=0), and try see where it fails. Perhaps an update should be nice...

**Comment by elieserdejesus** [comment]
Archlinux? I`m scared ... 😨  😄

**Comment by jonjamcam** [comment]
I see in the servers some use the camera with no problems. Others as soon as they enter and the remote video channel is being created in the mixer (not showing video, just loading the interval) they get kicked out. Rhen they try again a couple of times with same results and they give up.

I've been testing 2.1.11 today with good results.

I still get disconnections, but fewer, about 1 disconnected over 5 successful.

I have better results overall when turning on the camera before connecting to the server.

So we keep testing, we'll get to the bug eventually, it's just a matter of time and logging any observation.

**Comment by elieserdejesus** [comment]
@jonjamcam , if possible ask the "crashing users" about the log file. We need to know the OS and the camera resolution reported in the log file.

**Comment by jonjamcam** [comment]
@elieserdejesus  just tested latest commits from 2.1.12 beta. The limit in the camera upload stream is working great. My first tests are very promising. Fairly good image quality for such low bitrates and no disconnections ATM.

 I understand that the video limit is determined by the audio encoder value?

> @jonjamcam , if possible ask the "crashing users" about the log file

I've had no luck with that @elieserdejesus . New users that actually use the camera (very few) are more focused on jamming, but I keep asking :D

**Comment by elieserdejesus** [comment]
> I understand that the video limit is determined by the audio encoder value?

@jonjamcam, the video bit rate value is separated from audio. The possible values are in FFMpegMuxer.h:
https://github.com/elieserdejesus/JamTaba/blob/cb25e05b40970419d6ba49c1f3a077fe9a8a8a1f/src/Common/video/FFMpegMuxer.h#L37

And at moment the quality is hard coded to *Low*: https://github.com/elieserdejesus/JamTaba/blob/cb25e05b40970419d6ba49c1f3a077fe9a8a8a1f/src/Common/video/FFMpegMuxer.cpp#L86

**Comment by jonjamcam** [comment]
@elieserdejesus let's close this one for the moment. I think video is very stable now. If needed we can open it in the future.

**Comment by jonjamcam** [comment]
all good for cams now:

![image](https://user-images.githubusercontent.com/15310433/80664672-663fab00-8a65-11ea-9407-bec70e77cd84.png)

sharp image even with that low kbps. worth the waiting. I believe qt 5.6.3 was the game changer, but we'll never know I guess

**Comment by elieserdejesus** [comment]
very nice @jonjamcam 👍

**Comment by jonjamcam** [comment]
@Trashcanhead1111 commented in https://github.com/elieserdejesus/JamTaba/issues/1073#issuecomment-609007630

> Also everyone loved having the dual channels and the video same time .Can that be brought back like in 2.14 ? Jammer just asked if I could ask you as well.

Just for the record I spent some time today testing if the dual channel + video could work in 2.1.13 as now it's using channel 2. IMO channel functionality is very unpredictable when combining audio + video in the same channel. You get it working as  expected most of the time, but at some point bugs appear. For example after some time the transmit button ceased to have an effect on the other clients, or voice chat button pressed on channel 2 didn't activate the feature. This could be related to buggy QT versions, but to be safe I think ATM it should be avoided. Current behavior of separated channels works very well as far as I've been able to see in recent jams.
